File: apps/yukino/src/teams/task-stop.ts
/**
 * Copyright (c) 2026 hangtiancheng
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import type { TeamManager } from "./index.js";

import type { TaskManager } from "@/subagent/task-manager.js";
import type {
  Tool,
  ToolCategory,
  ToolContext,
  ToolResult,
  ToolSchema,
} from "@/tools/types.js";
import { strArg } from "@/utils/index.js";

/** Abort a running teammate or one-shot background Agent/Bash/PowerShell task. */
export class TaskStopTool implements Tool {
  name = "TaskStop";
  description =
    "Stop a running teammate or background task (Agent, Bash, PowerShell). Pass exactly one of teammate or task_id.";
  category: ToolCategory = "command";

  constructor(
    private teamManager: TeamManager,
    private taskManager?: TaskManager,
  ) {}

  schema(): ToolSchema {
    return {
      name: this.name,
      description: this.description,
      input_schema: {
        type: "object",
        properties: {
          teammate: {
            type: "string",
            description:
              "Name of the teammate to stop, exactly as it appears in the from= field of a task-notification",
          },
          task_id: {
            type: "string",
            description: "ID of a background task (Agent, Bash, PowerShell)",
          },
        },
      },
    };
  }

  async execute(
    ctx: ToolContext,
    args: Record<string, unknown>,
  ): Promise<ToolResult> {
    const name = strArg(args, "teammate", "");
    const taskId = strArg(args, "task_id", "");
    if ((!name && !taskId) || (name && taskId)) {
      return {
        output: "Error: pass exactly one of teammate or task_id",
        isError: true,
      };
    }

    if (taskId) {
      // The task registry of the loop running this call takes precedence: a
      // fork's background tasks live in its per-run manager, and task IDs are
      // per-manager counters, so the same ID in the constructor-injected
      // (host-level) manager may be a different task. Fall back to that
      // injected manager so a fork can still stop tasks it saw in its
      // pre-fork conversation snapshot.
      const manager = ctx.taskManager?.get(taskId)
        ? ctx.taskManager
        : this.taskManager;
      const task = manager?.get(taskId);
      if (!task) {
        return {
          output: `Error: background task '${taskId}' not found`,
          isError: true,
        };
      }
      if (task.status !== "running") {
        return {
          output: `Background task '${taskId}' is ${task.status}, nothing to stop`,
          isError: false,
        };
      }
      await manager?.stopAndWait(taskId);
      return { output: `Background task '${taskId}' stopped.`, isError: false };
    }

    // Teammate names may collide across teams; only stop within the team that actually has this member to avoid killing a namesake
    for (const team of this.teamManager.list()) {
      const member = team.members.get(name);
      if (!member) {
        continue;
      }

      if (!member.active) {
        return {
          output: `Teammate '${name}' in team '${team.name}' is not running, nothing to stop`,
          isError: false,
        };
      }
      await team.stopMember(name);
      return {
        output: `Teammate '${name}' in team '${team.name}' stopped.`,
        isError: false,
      };
    }

    return {
      output: `Error: teammate '${name}' not found. Known teammates: ${this.knownMembers()}`,
      isError: true,
    };
  }

  /** List all current teammate names for the model, so it doesn't keep retrying with a misremembered name */
  private knownMembers(): string {
    const names: string[] = [];
    for (const team of this.teamManager.list()) {
      for (const memberName of team.members.keys()) {
        names.push(memberName);
      }
    }
    return names.length > 0 ? names.join(", ") : "(none)";
  }
}

File: yukino/teams/task_stop.go
// Copyright (c) 2026 hangtiancheng
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

package teams

import (
	"context"
	"fmt"
	"strings"

	"github.com/hangtiancheng/yukino-code/yukino/tools"
)

// BackgroundTaskBoard abstracts the host's background-task registry so
// TaskStop can also stop task_id tasks (TS: TaskStopTool takes an optional
// TaskManager). teams cannot import the subagent package — subagent already
// imports teams — so the wiring layer adapts its TaskManager to this
// interface.
type BackgroundTaskBoard interface {
	// TaskState reports a background task's status ("pending", "running",
	// "completed", ...) and whether it exists.
	TaskState(id string) (state string, ok bool)
	// StopTask aborts a background task and reports whether it was stopped.
	StopTask(id string) bool
}

// TaskStopTool stops a running teammate or one-shot background task. Pass
// exactly one of teammate or task_id (TS: teams/task-stop.ts TaskStopTool).
//
// Teammates hang off the TeamManager (the Lead dispatches them via the Agent
// tool with team_name and the Team holds their Cancel functions); one-shot
// background tasks live on the BackgroundTaskBoard.
type TaskStopTool struct {
	TeamMgr *TeamManager
	// TaskBoard serves the task_id path; nil disables it.
	TaskBoard BackgroundTaskBoard
}

func (t *TaskStopTool) Name() string                 { return "TaskStop" }
func (t *TaskStopTool) Category() tools.ToolCategory { return tools.CategoryCommand }

func (t *TaskStopTool) Description() string {
	return "Stop a running teammate or background task (Agent, Bash, PowerShell). Pass exactly one of teammate or task_id."
}

func (t *TaskStopTool) Schema() map[string]any {
	return map[string]any{
		"name":        t.Name(),
		"description": t.Description(),
		"input_schema": map[string]any{
			"type": "object",
			"properties": map[string]any{
				"teammate": map[string]any{
					"type":        "string",
					"description": "Name of the teammate to stop, exactly as it appears in the from= field of a task-notification",
				},
				"task_id": map[string]any{
					"type":        "string",
					"description": "ID of a background task (Agent, Bash, PowerShell)",
				},
			},
		},
	}
}

func (t *TaskStopTool) Execute(ctx context.Context, args map[string]any) tools.ToolResult {
	name, _ := args["teammate"].(string)
	taskID, _ := args["task_id"].(string)
	if (name == "" && taskID == "") || (name != "" && taskID != "") {
		return tools.ToolResult{Output: "Error: pass exactly one of teammate or task_id", IsError: true}
	}

	if taskID != "" {
		// TS task-stop.ts: the task registry of the loop running this call
		// takes precedence — task ids are per-manager counters, so a fork's
		// background tasks live in its per-run manager and the same id on the
		// host-level board may be a different task. Fall back to the injected
		// board so a fork can still stop tasks from its pre-fork snapshot.
		board := t.TaskBoard
		if mgr, ok := tools.BackgroundTaskManagerFromContext(ctx); ok && mgr != nil {
			if b, isBoard := mgr.(BackgroundTaskBoard); isBoard {
				if _, exists := b.TaskState(taskID); exists {
					board = b
				}
			}
		}
		if board == nil {
			return tools.ToolResult{
				Output:  fmt.Sprintf("Error: background task '%s' not found", taskID),
				IsError: true,
			}
		}
		state, ok := board.TaskState(taskID)
		if !ok {
			return tools.ToolResult{
				Output:  fmt.Sprintf("Error: background task '%s' not found", taskID),
				IsError: true,
			}
		}
		if state != "running" {
			return tools.ToolResult{
				Output: fmt.Sprintf("Background task '%s' is %s, nothing to stop", taskID, state),
			}
		}
		board.StopTask(taskID)
		return tools.ToolResult{Output: fmt.Sprintf("Background task '%s' stopped.", taskID)}
	}

	if t.TeamMgr == nil {
		return tools.ToolResult{Output: "Error: team manager unavailable", IsError: true}
	}

	// Teammate names may be duplicated across teams; only stop in the team
	// that actually contains this member to avoid killing a same-named
	// teammate in another team.
	for _, teamName := range t.TeamMgr.ListTeams() {
		team := t.TeamMgr.GetTeam(teamName)
		if team == nil {
			continue
		}
		if !team.HasMember(name) {
			continue
		}
		if !team.IsMemberActive(name) {
			return tools.ToolResult{
				Output: fmt.Sprintf("Teammate '%s' in team '%s' is not running, nothing to stop", name, teamName),
			}
		}
		team.StopMember(name)
		return tools.ToolResult{
			Output: fmt.Sprintf("Teammate '%s' in team '%s' stopped.", name, teamName),
		}
	}

	return tools.ToolResult{
		Output:  fmt.Sprintf("Error: teammate '%s' not found. Known teammates: %s", name, t.knownMembers()),
		IsError: true,
	}
}

// knownMembers lists all current teammate names for the model, preventing it
// from retrying endlessly with a misremembered name.
func (t *TaskStopTool) knownMembers() string {
	var names []string
	for _, teamName := range t.TeamMgr.ListTeams() {
		team := t.TeamMgr.GetTeam(teamName)
		if team == nil {
			continue
		}
		names = append(names, team.MemberNames()...)
	}
	if len(names) == 0 {
		return "(none)"
	}
	return strings.Join(names, ", ")
}
